crates/ra_analysis/src/descriptors/mod.rs

```diff
@@ -31,17 +36,86 @@ impl FnDescriptor {
             node.syntax().text().to_string()
         };
 
+        if let Some((comment_range, docs)) = FnDescriptor::extract_doc_comments_for_fndef(node) {
+            let comment_range = comment_range.checked_sub(node.syntax().range().start()).unwrap();
+            let start = u32::from(comment_range.start()) as usize;
+            let end = u32::from(comment_range.end()) as usize;
+
+            label.replace_range(start..end, "");
+
+            // Massage markdown
+            let mut processed_lines = Vec::new();
+            let mut in_code_block = false;
+            for line in docs {
+                if line.starts_with("```") {
+                    in_code_block = !in_code_block;
+                }
+
+                let line = if in_code_block && line.starts_with("```") && !line.contains("rust") {
+                    "```rust".into()
+                } else {
+                    line.to_string()
+                };
+
+                processed_lines.push(line);
+            }
+
+            if !processed_lines.is_empty() {
+                doc = Some(processed_lines.join("\n"));
+            }
+        }
+
         let params = FnDescriptor::param_list(node);
         let ret_type = node.ret_type().map(|r| r.syntax().text().to_string());
 
         Some(FnDescriptor {
             name,
             ret_type,
             params,
-            label,
+            label: label.trim().to_owned(),
+            doc
         })
     }
 
+    fn extract_doc_comments_for_fndef(node: ast::FnDef) -> Option<(TextRange, Vec<String>)> {
```

Comment:
I think some part of this functionality should leave in the ra_syntax crate.

Like this

```
trait DocCommentsOwner {
    fn doc_comments(self) -> AstChildren<DocCOmment>;
}
```

Reply:
That simplifies things